### ISSUE TITLE
Standardize pattern for exposing advanced configuration for compression streams

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#42820]
There is [a desire](https://github.com/dotnet/runtime/issues/2236#issuecomment-699927812) to expose compression-algorithm specific configuration when constructing compression streams. The same options can be used for `ZLibStream`, `DeflateStream`, `GZipStream`. Named those options with `Zlib` prefix same as the native library name they use another prefix we considered is `Deflate` - the algorithm name, the prefix needed because we have existing `CompressionLevel` enum that used for all compression streams including `BrotliStream` and it cannot be used for fine tuning the compression level for specific stream
https://github.com/dotnet/runtime/blob/d710379da9b2699bba0a8f26d65d3a6e8549f26a/src/libraries/System.IO.Compression/ref/System.IO.Compression.cs#L9-L15 

### API Proposal
```C#
namespace System.IO.Compression
{
    public enum ZlibCompressionLevel
    {
        DefaultCompression = -1,
        NoCompression = 0,
        Level1 = 1,
        BestSpeed = Level1,
        Level2 = 2,
        Level3 = 3,
        Level4 = 4,
        Level5 = 5,
        Level6 = 6,
        Level7 = 7,
        Level8 = 8,
        Level9 = 9,
        BestCompression = Level9,
    }
    public enum ZlibCompressionStrategy
    {
        DefaultStrategy = 0,
        Filtered = 1,
        HuffmanOnly = 2,
        Rle = 3,
        Fixed = 4
    }
    public enum ZlibFlushMode
    {
        NoFlush = 0,
        PartialFlush = 1,
        SyncFlush = 2,
        FullFlush = 3,
        Finish = 4,
    }
    public sealed class ZLibStream : Stream
    {
        // CompressionMode.Compress
        public ZLibStream(Stream stream, ZlibCompressionLevel compressionLevel = ZlibCompressionLevel.DefaultCompression,
            ZlibCompressionStrategy strategy = ZlibCompressionStrategy.DefaultStrategy, bool leaveOpen = false);
        public ZlibFlushMode FlushMode { get; set; }
    }
    public partial class DeflateStream : Stream
    {
        public DeflateStream(Stream stream, ZlibCompressionLevel compressionLevel = ZlibCompressionLevel.DefaultCompression,
            ZlibCompressionStrategy strategy = ZlibCompressionStrategy.DefaultStrategy, bool leaveOpen = false);
        public ZlibFlushMode FlushMode { get; set; }
     }
     public partial class GZipStream : Stream
    {
        public GZipStream(Stream stream, ZlibCompressionLevel compressionLevel = ZlibCompressionLevel.DefaultCompression,
            ZlibCompressionStrategy strategy = ZlibCompressionStrategy.DefaultStrategy, bool leaveOpen = false) { }
        public ZlibFlushMode FlushMode { get; set; }
    }
    public enum BrotliCompressionQuality
    {
        NoCompression,
        Quality1,
        Quality2,
        Quality3,
        Quality4,
        Quality5,
        Quality6,
        Quality7,
        Quality8,
        Quality9,
        Quality10,
        Quality11
    }
    public sealed partial class BrotliStream : System.IO.Stream
    {
        public BrotliStream(Stream stream, BrotliCompressionQuality quality = BrotliCompressionQuality.Quality4, bool leaveOpen = false) { }
    }
}
```
`ZlibCompressionLevel`, `ZlibCompressionStrategy` are only for Compress mode.

Currently `ZlibFlushMode.NoFlush` used in normal `Read/Write` operations, `ZlibFlushMode.SyncFlush ` used for Stream.Flush  and `ZlibFlushMode.Finish` is used on dispose. The value set by the new `FlushMode` property will be used for normal `Read/Write` operations only.

The `MemoryLevel` and/or `WindowBits` options are omitted because there is no ask for them, we could add these and other options if/when they are requested. 

### API Usage

```cs
private MemoryStream CompressStream(Stream uncompressedStream)
{
    var compressorOutput = new MemoryStream();
    using (var compressionStream = new ZLibStream(compressorOutput, compressionLevel: ZlibCompressionLevel.Level5, strategy: ZlibCompressionStrategy.Filtered, leaveOpen: true))
    {
        compressionStream.FlushMode = ZlibFlushMode.NoFlush;
        var buffer = new byte[4096];
        int bytesRead;
        while ((bytesRead = uncompressedStream.Read(buffer, 0, buffer.Length)) > 0)
        {
            compressionStream.Write(buffer, 0, bytesRead);
        }
    }

    compressorOutput.Position = 0;
    return compressorOutput;
}
```

### Alternative Designs
An alternative design would introduce an options type, maybe per each stream, like so:

```C#
namespace System.IO.Compression
{
    public class BrotliOptions
    {
        public readonly BrotliCompressionQualityQuality { get; }
        public readonly int WindowBits { get; }
        public BrotliOptions(BrotliCompressionQuality quality = BrotliCompressionQuality.Level4, int windowBits = 22) { }
    }
    public sealed class BrotliStream : Stream
    {
        public BrotliStream(Stream stream, BrotliOptions options, bool leaveOpen = false);
    }
    public struct ZLibOptions
    {
        public int WindowBits { get; }
        public ZlibMemoryLevel MemoryLevel { get; }
        public int CompressionLevel { get; }
        public int CompressionStrategy { get; }
        public ZlibOptions(ZlibCompressionLevel compressionLevel = ZlibCompressionLevel.DefaultCompression, 
             ZlibCompressionStrategy strategy = ZlibCompressionStrategy.DefaultStrategy,
             ZlibMemoryLevel memoryLevel = ZlibMemoryLevel.Level8, int windowBits = -1);
    }
    public sealed class ZLibStream : Stream
    {
        public ZLibStream(Stream stream, ZLibOptions options, bool leaveOpen = false);
    }
    public class DeflateStream : Stream
    {
        public DeflateStream(Stream stream, ZLibOptions options, bool leaveOpen = false);
    }
    public class GZipStream : Stream
    {
        public GZipStream(Stream stream, ZLibOptionsoptions, bool leaveOpen = false);
    }
}
```

We can then decide whether we want to be in the business of defining enums for the underlying types or whether we consider them passthru.

**Comment by FiniteReality** [comment]
I personally like the `Options` approach; it's easier to pass around as part of configuration options, and IMO makes refactoring (e.g. to change the compression algorithm being used) easier if that becomes an issue.

**Comment by JimBobSquarePants** [comment]
The `Options` approach doesn't allow for setting the `ZlibFlushMode` property for each read/write. That was a requirement identified in the previous thread.

**Comment by FiniteReality** [comment]
> The `Options` approach doesn't allow for setting the `ZlibFlushMode` property for each read/write. That was a requirement identified in the previous thread.

This isn't necessarily true:
```cs
class ZLibOptions
{
    public static readonly int FlushFinish = 4;
    public int FlushMode { get; set; }
}

class ZLibStream : Stream
{
    public ZLibStream(Stream stream, ZLibOptions options, bool leaveOpen = false);
    public ZLibOptions Options { get; set; }
}
// ...
using var stream = new ZLibStream(myStream, options);
stream.Options.FlushMode = ZLibOptions.FlushFinish;
stream.Write(...);
```

However, couldn't the existing `Flush` API be augmented to allow you to specify the flush mode? I imagine, by default, you'd want `Z_NO_FLUSH` and then would want to specify something special in a separate step. I'm not sure if ZLib supports that sort of behaviour though

**Comment by JimBobSquarePants** [comment]
Ah yeah, no. Deflate.Flush is a separate operation from Stream.Flush.

**Comment by FyiurAmron** [comment]
FWIW, I'd say that current abstraction is broken (both in terms of correspondence with e.g. actual zlib functions, of POLA, and in simplicity of use) and adding a better abstraction would be a better idea. https://github.com/dotnet/runtime/issues/2236#issuecomment-698981107 sums up my stance on this. As such, I'd say that providing an access to more low-level API would solve all the issues raised here (and in #2236, #16923, and probably some other places too).

What am I talking about precisely? Look at actual zlib. We of course can skip both the init/end as separate calls, but still: both deflate() and inflate() calls take `z_stream_s` & `flush` as param, which can be roughly translated to 3 OOP params: input stream, output stream, and call options. While some options are arguably useless in C# (alloc fn, dealloc fn), the architecture would be still the same. zlib provides two explicit methods for this. Using a single class for two different (albeit related) functionalities is akin to having a serializer being able to deserialize etc. In this case, the semantics are outright ugly:

```
DeflateStream compressStream = new( srcStream, CompressionLevel.Optimal );
DeflateStream decompressStream = new( srcStream, CompressionMode.Decompress );
```

This is neither intuitive or useful in any way. A proper way would be to have either:

```
DeflateStream compressStream = new( srcStream, options );
InflateStream decompressStream = new( srcStream, options );
```

or, if for some reason we want to have them bundled into a single class (I would advocate against it, but it would be still better than current approach):

```
ZlibStream compressStream = new( srcStream, Mode.Compress, options );
ZlibStream decompressStream = new( srcStream, Mode.Decompress, options );
```

Using an option pattern for flushing would be IMVHO the best option (sic!), since we can just set flush option as read-write (as @FiniteReality stated). I would keep the other options, especially those used in `init` or related to it (mode [compress/decompress], compression level, compression strategy, data type etc.) read-only. This way, we can have the best of all - access to low-level operations, easy backwards-compatible extensibility, clear abstractions. Switching the used headers etc.  would be trivial this way.

**Comment by rog1039** [comment]
I've landed here because the current API for managing Brotli compression only gives me:

```c#
public enum CompressionLevel
{
        Optimal = 0,
        Fastest = 1,
        NoCompression = 2
}
```
In one sample, optimal compressed to 700K in 53sec and fastest to 4,400K in 2.8sec. I'd like to find a middle ground between these two extremes for our application. 

I'm not too hung up on how we end up tweaking the various knobs but do think it is important to have access to all (or near all) the knobs for a given compression algorithm.

From a cursory reading of this and related threads, I get the feeling that a dotnet-wide compression enum is an abstraction that is trying to cover too many use cases. The options style approach seems fine to me.

**UPDATE**
I just went through a blog post that suggested I do
```
Level = (CompressionLevel)5;
```
as a way that may effect the compression level. This did work but it was not clear how or why this would do what I wanted.  It was only after going through dotnet source code that it made any sense. First to the BrotliStream:
![image](https://user-images.githubusercontent.com/5702706/103809446-84fc3580-5027-11eb-945b-4321e2ccc631.png)

and then into 
![image](https://user-images.githubusercontent.com/5702706/103809358-64cc7680-5027-11eb-8ca3-3ec51e8406fb.png)

So, in summary, I would have been saved a lot of headache and time if we simply exposed algorithm specific options to the programmer. The current behavior is unintuitive and limiting. A new options-based configuration API sounds expressive and intuitive.

**Comment by terrajobst** [comment]
@carlossanlop any objections marking this as api-ready-for-review? We can discuss this in the meeting if we want to do this at all.

**Comment by rog1039** [comment]
I think a new Api could be worthwhile. Just a few minutes ago I ran into another problem with the current API design. I upgraded a WebApi project to .Net 6 and some of my endpoint response times dropped from 200ms to 3seconds. I was very confused why my response time was 15x longer than before. Turns out it was due to the design of the compression API's. 

I used 
```csharp
services.AddResponseCompression(options =>
            {
                options.MimeTypes = CompressedMimeTypes;
                options.Providers.Add(
                    new BrotliCompressionProvider(new BrotliCompressionProviderOptions()
                    {
                        Level = (CompressionLevel) 3
                    }));
            });
```
in Net5 to get a brotli compression level of 3 which provided a good combination of compression size vs compression time. 

However, in Net6, the CompressionLevel enum was expanded to: 
```csharp
CompressionLevel
  {
    Optimal,
    Fastest,
    NoCompression,
    SmallestSize,
  }
```
Now, my code instead of using Brotli compression level 3 was using level 11!

Why was that? Because of this code:
```csharp
 internal static int GetQualityFromCompressionLevel(CompressionLevel level) =>
            level switch
            {
                CompressionLevel.Optimal => Quality_Default,
                CompressionLevel.NoCompression => Quality_Min,
                CompressionLevel.Fastest => 1,
                CompressionLevel.SmallestSize => Quality_Max,
                _ => (int)level,
            };
```

The ```CompressionLevel``` enum has 4 members now so Brotli compression levels 1, 2, and 3 are not even usable because they map to enum members. Fortunately, brotli compression level 4 has compression and runtimes similar to level 3 so it did not matter in our case much.

The more I think about this issue, the more I lean towards a new API for compression. So another +1 for a new options-based or some other new API where compression API's are explicit and understandable rather than the confusing situation we have today.

**Comment by rog1039** [comment]
Further, the choice of enum conversions in BrotliUtils.cs is strange. Optimal compression and SmallestSize both map to the highest compression, value 11.

Smallest size makes sense as 11, however, Optimal should not IMHO map to 11. In my application, on a roughly 1.5MB payload, compression level 3 reduces the payload by 95% and takes 100ms. Compression level 11 reduces the payload by 97% and takes 3,000ms. Doesn't seem like it is worth taking 15 times as long for an additional 2% reduction in size. Optimal in BrotliUtils.cs should map to something else like level 3. Better still of course is a new API, but this seems like a quick 1-line change that would make Brotli easier to use by avoiding the probably never used compression level 11.

**Comment by AraHaan** [comment]
For zlib, I got some progress on:

For my local branch of dotnet/runtime I have the following added to System.IO.Compression:

```diff
namespace System.IO.Compression
{
+    public enum ZlibCompressionLevel
+    {
+        DefaultCompression = -1,
+        NoCompression,
+        Level1,
+        BestSpeed = Level1,
+        Level2,
+        Level3,
+        Level4,
+        Level5,
+        Level6,
+        Level7,
+        Level8,
+        Level9,
+        BestCompression = Level9
+    }
+    public enum ZlibCompressionStrategy
+    {
+        DefaultStrategy,
+        Filtered,
+        HuffmanOnly,
+        Rle,
+        Fixed
+    }
+    public enum ZlibFlushCode
+    {
+        NoFlush,
+        PartialFlush,
+        SyncFlush,
+        FullFlush,
+        Finish,
+        Block,
+        Trees
+    }
+    public struct ZlibOptions
+    {
+        public int WindowBits { get { throw null; } set { throw null; } }
+        public int MemoryLevel { get { throw null; } set { throw null; } }
+        public ZlibFlushCode FlushMode { get { throw null; } set { throw null; } }
+        public ZlibCompressionLevel CompressionLevel { get { throw null; } set { throw null; } }
+        public ZlibCompressionStrategy Strategy { get { throw null; } set { throw null; } }
+        public CompressionMode CompressionMode { get { throw null; } }
+        public ZlibOptions() { }
+        public ZlibOptions(ZlibCompressionLevel compressionLevel) { }
+    }
```

I also have changes to add an public ZlibEncoder and ZlibDecoder struct that can be publicly exposed to end users and have the stream implementations switch to use only the encoder / decoder to zlib as well since they have similar code (only simplified and cleaned up somewhat).

However, as much as it pains me to say, I think the stream based compression types should be deprecated for an Encoder/Decoder-like implementation.

cc @terrajobst can I PR the changes I done to Zlib since I have to propose the encoder and decoder (another issue soon) that will depend on it existing in main.

As for ZLibStream and DeflateStream (as well as the other streams for Compression except for Zip* based types that are stream-like), we need to consider if we should mark those as obsolete so people can migrate to span/memory based compression apis that are encoder/decoder like and feel more natural.

**Comment by stephentoub** [comment]
> we need to consider if we should mark those as obsolete so people can migrate to span/memory based compression apis that are encoder/decoder like and feel more natural.

No, we will not be obsoleting/deprecating the Stream-based compression APIs.

**Comment by AraHaan** [comment]
> > we need to consider if we should mark those as obsolete so people can migrate to span/memory based compression apis that are encoder/decoder like and feel more natural.
> 
> No, we will not be obsoleting/deprecating the Stream-based compression APIs.

Alright, anyways does this zlib part LGTM? https://github.com/AraHaan/runtime/commit/5f7ea7eb8ce2f8ca83cb6dfedda4c278a750d0df

**Comment by AraHaan** [comment]
I have edited this comment into an api proposal: https://github.com/dotnet/runtime/issues/42820#issuecomment-980830626

**Comment by masonwheeler** [comment]
I just downloaded the .NET 7 RC.  I'm no longer able to select a BrotliStream compression level manually by casting an int to `CompressionLevel`, but there's no replacement API such as this to provide that functionality in a more standardized way.  Wasn't this supposed to end up in .NET 7 as a part of the package so that the valuable ability to fine-tune compression didn't get pulled out from under existing codebases?

**Comment by AraHaan** [comment]
It sadly did not get reviewed in time to land in .NET 7.

**Comment by andtii** [comment]
It is pretty bad that you removed the support of setting any compression level supported. When updating to .net 7 we started getting exceptions :(
![image](https://user-images.githubusercontent.com/1434571/208057094-8e7a5f5d-3107-4980-bf3a-a4d9cf988b64.png)

**Comment by AraHaan** [comment]
The encoders will be a new normal for when you want to set custom compression levels not covered by the current CompressionLevel enum.

As for the proposed zlib encoder / decoder based on spans, I hope it will land in .NET 8.

**Comment by masonwheeler** [comment]
> As for the brotlistream, I thought there was an argument to pass in an int compression level into it

Look up about 3 posts.  There was going to be one, but the feature got removed due to time running out for testing it properly.  And yet they still put in the breaking change that it was meant to alleviate, so now apparently life sucks for everyone until .NET 8.

**Comment by AraHaan** [comment]
I made an example API lib that people can try out today and supports .NET 8. https://github.com/Elskom/ZlibSharp

Note I have not updated it yet from 1.2.13 of zlib. If you want could pull request it in, but I would need to construct a nuget package and push the updated native libs that is a dependency of that.

**Comment by buyaa-n** [comment]
Thank you @AraHaan for API proposal, I made a few changes like:
- Moved `ZlibFlushMode` property from `ZlibOptions` into the `*Stream` classes as it needs to be there
- Do not think we really need `ZlibOptions` for each Stream, might want to rename `ZlibOptions` into more generic one
- Removed `CompressionMode` from `ZlibOptions`, don see its needed to be there
- Renamed `BrotliCompressionLevel` into `Quality`, options only apply for `CompressionMode.Compress` mode

I wonder do we really need those getter properties in `ZlibOptions`? It appears to me that they only need for initializing the deflate streams. Is there any use case for needing those options after init? If not, we could remove `ZlibOptions`, `BrotliOptions`

Here is modified API Proposal:

## API Design
```diff
namespace System.IO.Compression
{
+    public enum ZlibCompressionLevel 
+    {
+        DefaultCompression = -1,
+        NoCompression,
+        Level1,
+        BestSpeed = Level1,
+        Level2,
+        Level3,
+        Level4,
+        Level5,
+        Level6,
+        Level7,
+        Level8,
+        Level9,
+        BestCompression = Level9
+    }
+    public enum ZlibCompressionStrategy
+    {
+        DefaultStrategy,
+        Filtered,
+        HuffmanOnly,
+        Rle,
+        Fixed
+    }
+    public enum ZlibFlushMode
+    {
+        NoFlush,
+        PartialFlush,
+        SyncFlush,
+        FullFlush,
+        Finish,
+        Block,
+        Trees
+    }
+    public enum ZlibMemoryLevel
+    {
+        Level1 = 1,
+        Level2,
+        Level3,
+        Level4,
+        Level5,
+        Level6,
+        Level7,
+        Level8,
+        Level9
+    }
+    public class ZlibOptions
+    {
+        public int WindowBits { get { throw null; } }
+        public ZlibMemoryLevel MemoryLevel { get { throw null; } }
+        public ZlibCompressionLevel CompressionLevel { get { throw null; } }
+        public ZlibCompressionStrategy Strategy { get { throw null; } }
+        public ZlibOptions(ZlibCompressionLevel compressionLevel = ZlibCompressionLevel.DefaultCompression, ZlibCompressionStrategy strategy = ZlibCompressionStrategy.DefaultStrategy, ZlibMemoryLevel memoryLevel = ZlibMemoryLevel.Level1, int windowBits = -1) { }
+    }

    public sealed partial class ZLibStream : Stream
    {
+        public ZlibStream(Stream stream, ZlibOptions zlibOptions, CompressionMode compressionMode = CompressionMode.Compress, bool leaveOpen = false) { }
+        public ZlibFlushMode FlushMode { get; set; }
    }
    public sealed partial class GZipStream : Stream
    {
+        public GZipStream(Stream stream, ZlibOptions zlibOptions, CompressionMode compressionMode = CompressionMode.Compress, bool leaveOpen = false) { }
+        public ZlibFlushMode FlushMode { get; set; }
    }
    public sealed partial class DeflateStream : Stream
    {
+        public DeflateStream(Stream stream, ZlibOptions zlibOptions, CompressionMode compressionMode = CompressionMode.Compress, bool leaveOpen = false) { }
+        public ZlibFlushMode FlushMode { get; set; }
    }
+    public enum Quality
+    {
+        NoCompression,
+        Level1,
+        Level2,
+        Level3,
+        Level4,
+        Level5,
+        Level6,
+        Level7,
+        Level8,
+        Level9,
+        Level10,
+        Level11
+    }
+    public struct BrotliOptions
+    {
+        public readonly Quality Quality { get { throw null; } }
+        public readonly int WindowBits { get { throw null; } }
+        public BrotliOptions(Quality quality = Quality.Level4, int windowBits = 22) { }
+    }
    public sealed partial class BrotliStream : System.IO.Stream
    {
+        public BrotliStream(Stream stream, BrotliOptions options, bool leaveOpen = false) { }
    }
```

## Alternate Design

- Just use the enums in the `*Strem` constructors, directly no need `ZlibOptuons` or `BrotliOptions`
- Don't add enums, just use int for each option for new `*Strem` constructors

CC @terrajobst @stephentoub

**Comment by julealgon** [comment]
> It is pretty bad that you removed the support of setting any compression level supported. When updating to .net 7 we started getting exceptions :( @terrajobst ![image](https://user-images.githubusercontent.com/1434571/208057094-8e7a5f5d-3107-4980-bf3a-a4d9cf988b64.png)

Shouldn't this breaking change be reversed? As it stands, I'd consider it a bug.

Tying that change to this entire proposal here is a mistake IMHO, as this is much broader and could end up just delaying that small piece of work by not being approved etc.

Or does the discussion on #95460 basically make it impossible to reverse properly now @stephentoub ?

**Comment by buyaa-n** [comment]
> Shouldn't this breaking change be reversed? As it stands, I'd consider it a bug.

I believe that is answered in https://github.com/dotnet/runtime/issues/72220

**Comment by stephentoub** [comment]
> Shouldn't this breaking change be reversed?

No. It was a bug it was allowed in the first place.

**Comment by buyaa-n** [comment]
@JimBobSquarePants @AraHaan and whoever interested in these APIs, do you need the `MemoryLevel` and/or `windowBits` options exposed for ZLib? Also, is anyone needs the `windowBits` option exposed for Brotli? I looked around a bit here and other related issues/PRs don't see anyone mentioned that they need these options, so I would like to remove those options from the proposal.

**Comment by JimBobSquarePants** [comment]
@buyaa-n I don’t think I would ever have a need for them and I don’t think many people would. Compression levels and type are commonly used features

**Comment by masonwheeler** [comment]
> > Shouldn't this breaking change be reversed?
> 
> No. It was a bug it was allowed in the first place.

"Bug" or not, it's still a breaking change because people wrote code that depends on it, that code worked as expected, and it no longer works in the updated version.  That's a real problem, one that Microsoft [used to understand well.](https://www.joelonsoftware.com/2004/06/13/how-microsoft-lost-the-api-war/)

**Comment by stephentoub** [comment]
> it's still a breaking change

And was documented as such.
https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/7.0/brotlistream-ctor

**Comment by AraHaan** [comment]
> @buyaa-n I don’t think I would ever have a need for them and I don’t think many people would. Compression levels and strategy are commonly used features.

While that may be true, there are times where people want to make software using custom window bits so having a public api that supports providing it is a free win for them.

**Comment by AraHaan** [comment]
And here is an image explaining why to expose window bits at least for zlib:

![image](https://github.com/user-attachments/assets/4652ef0b-6c2e-4805-8cb7-303c87f075ff)

In my ``ZlibSharp`` package, I exposed it because it is pretty common for an application to want to compress items using ``Deflate``, ``zlib-streams``, and ``gzip streams``. An example software that comes in mind is https://www.github.com/python/cpython/ because their built-in standard library has a hard requirement to support all 3 types. Also my ``ZlibSharp`` package I have finally decided to apply my ``zlib encoder`` and ``zlib decoder`` logic to it to give people a preview of that and the type of ABI surface I want for when I introduce it as ``System.IO.Compression.Zlib`` which would be a better designed implementation.

**Comment by stephentoub** [comment]
> it is pretty common for an application to want to compress items using Deflate, zlib-streams, and gzip streams

For which DeflateStream, ZLibStream, and GZipStream exist.

**Comment by AraHaan** [comment]
> > it is pretty common for an application to want to compress items using Deflate, zlib-streams, and gzip streams
> 
> For which DeflateStream, ZLibStream, and GZipStream exist.

I sorta made something that can replace those streams via ``ZlibEncoder`` and ``ZlibDecoder`` and will arrive in my ``ZlipSharp`` preview package that can be tested as a proof of concept for a new ``System.IO.Compression.Zlib`` assembly in the runtime.

**Comment by buyaa-n** [comment]
> > > it is pretty common for an application to want to compress items using Deflate, zlib-streams, and gzip streams
> > 
> > 
> > For which DeflateStream, ZLibStream, and GZipStream exist.
> 
> I sorta made something that can replace those streams via `ZlibEncoder` and `ZlibDecoder` and will arrive in my `ZlipSharp` preview package that can be tested as a proof of concept for a new `System.IO.Compression.Zlib` assembly in the runtime.

That doesn't sound like a reason to add those for runtime.

**Comment by AraHaan** [comment]
> > > > it is pretty common for an application to want to compress items using Deflate, zlib-streams, and gzip streams
> > > 
> > > 
> > > For which DeflateStream, ZLibStream, and GZipStream exist.
> > 
> > 
> > I sorta made something that can replace those streams via `ZlibEncoder` and `ZlibDecoder` and will arrive in my `ZlipSharp` preview package that can be tested as a proof of concept for a new `System.IO.Compression.Zlib` assembly in the runtime.
> 
> That doesn't sound like a reason to add those for runtime. We can add that later, when it's really needed. I removed them from the API proposal.

I disagree, having only stream based API's is counter productive if they are in hot path code as it would reduce performance. Span based on the other hand would increase performance of those hot paths and can make a difference in places.

**Comment by stephentoub** [comment]
>> That doesn't sound like a reason to add those for runtime. We can add that later, when it's really needed. I removed them from the API proposal.

> I disagree, having only stream based API's is counter productive

The two of you are talking about different things. @buyaa-n is taking about Stream ctors that take the additional settings. @AraHaan is talking about non-Stream-based encoder/decoders. This issue is about Stream ctors.

**Comment by buyaa-n** [comment]
> > > That doesn't sound like a reason to add those for runtime. We can add that later, when it's really needed. I removed them from the API proposal.
> 
> > I disagree, having only stream based API's is counter productive
> 
> The two of you are talking about different things. @buyaa-n is taking about Stream ctors that take the additional settings. @AraHaan is talking about non-Stream-based encoder/decoders. This issue is about Stream ctors.

Right, sorry for the confusion, I am talking about adding `windowBits` option for `DeflateStream`, `ZLibStream`, and `GZipStream` (and `Brotli`). Don't see its needed for these streams now, and we can add it later when needed.

**Comment by bartonjs** [comment]
* ZlibCompressionStrategy => ZlibCompressionStrategy (consistent casing with ZLibStream)
* ZlibCompressionStrategy.Rle => RunLengthEncoding
* ZlibCompressionStrategy.DefaultStrategy => Default
* ZLibFlushMode is cut pending a scenario
* ZLibCompressionLevel is cut, just using the raw integer.
* Created ZLibCompressionOptions to allow for expanding to more options in the future
* ZLibCompressionOptions is a required parameter, consistent with the existing overloads.
* BrotliCompressionQuality => int property on a new class

```C#
namespace System.IO.Compression
{
    public enum ZLibCompressionStrategy
    {
        Default = 0,
        Filtered = 1,
        HuffmanOnly = 2,
        RunLengthEncoding = 3,
        Fixed = 4,
    }
    public sealed class ZLibCompressionOptions
    {
        public int CompressionLevel { get; set; }
        public ZLibCompressionStrategy CompressionStrategy { get; set; }
    }
    public sealed class ZLibStream : Stream
    {
        // CompressionMode.Compress
        public ZLibStream(Stream stream, ZLibCompressionOptions compressionOptions, bool leaveOpen = false);
    }
    public partial class DeflateStream : Stream
    {
        public DeflateStream(Stream stream, ZLibCompressionOptions compressionOptions, bool leaveOpen = false);
    }
    public partial class GZipStream : Stream
    {
        public GZipStream(Stream stream, ZLibCompressionOptions compressionOptions, bool leaveOpen = false);
    }


    public sealed class BrotliCompressionOptions
    {
        public int Quality { get; set; }
    }
    public sealed partial class BrotliStream : System.IO.Stream
    {
        public BrotliStream(Stream stream, BrotliCompressionOptions compressionOptions, bool leaveOpen = false) { }
    }
}
```

**Comment by Mrgaton** [comment]
What about dictionary size or window length

**Comment by Mrgaton** [comment]
Compresión level maybe would be better with an enum to know which options are available

**Comment by Mrgaton** [comment]
I think node js can compress brotli with multiple threads, could this be possible on c# too?

**Comment by buyaa-n** [comment]
> What about dictionary size or window length

If you need them, please open a new issue (API poposal) with your user scenario, we are open to add more options to the new `***CompressionOptions` types

**Comment by Mrgaton** [comment]
I tried opening them and I couldn't do it If you can do it, please

I would like to be able to increase dictionary length and set custom brotli mode

**Comment by stephentoub** [comment]
> I tried opening them and I couldn't do it If you can do it, please

What prevented you from doing so?

**Comment by Mrgaton** [comment]
I didn't have permission or something it gives me an error

**Comment by masonwheeler** [comment]
If you could copy/paste the exact text of the error here, that would be extremely helpful.

**Comment by buyaa-n** [comment]
> I didn't have permission or something it gives me an error

No permission needed for creating an issue
